Add SubCircleDomain and SparseSubCircleDomain

atgrosso opened the pull request:
Depends on #100

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101)
<!-- Reviewable:end -->

andrewmilson reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101)* status: 0 of 11 files reviewed, 4 unresolved discussions (waiting on @atgrosso)

___
*[`stwo_cairo_verifier/src/queries.cairo` line 30 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101#-O8EyLf69LBWKNM6WM_9:-O8EyLf69LBWKNM6WM_A:b-3dufzo) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/22e3e84e17d76d240d9ca0167261d7f48f92621d/stwo_cairo_verifier/src/queries.cairo#L30)):*
> ```Cairo
>             i = i + 1;
>         };
>         res
> ```

I think its better for readability using for loops vs while loops where appropriate.



_Suggestion:_
```Cairo
        let sub_circle_size = pow(2, *self.log_size);
        let start = *self.coset_index * sub_circle_size;
        let end = start + sub_circle_size;
        let mut res = array![];
        for i in start..end {
            res.append(i);
        };
        res
```
___
*[`stwo_cairo_verifier/src/queries.cairo` line 51 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101#-O8F-xIUB2txPuljeEnG:-O8F-xIUB2txPuljeEnH:bj3lh64) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/22e3e84e17d76d240d9ca0167261d7f48f92621d/stwo_cairo_verifier/src/queries.cairo#L51)):*
> ```Cairo
> #[generate_trait]
> pub impl SparseSubCircleDomainImpl of SparseSubCircleDomainTrait {
>     fn flatten(self: @SparseSubCircleDomain) -> Span<usize> {
> ```

Why not return Array?



_Code quote:_
```Cairo
Span<usize>
```
___
*[`stwo_cairo_verifier/src/queries.cairo` line 63 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101#-O8F-zpW2xkqLcoXP7az:-O8F-zpW2xkqLcoXP7b-:b-t6gsf6) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/22e3e84e17d76d240d9ca0167261d7f48f92621d/stwo_cairo_verifier/src/queries.cairo#L63)):*
> ```Cairo
>             i = i + 1;
>         };
>         res.span()
> ```

Same here perhaps



_Suggestion:_
```Cairo
        let mut res = array![];
        for domain in self.domains.span() {
            for position in domain.to_decommitment_positions() {
                res.append(position);
            };
        };
        res
```
___
*[`stwo_cairo_verifier/src/queries.cairo` line 68 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101#-O8Exk5I0Vm79AX6IenH:-O8Exk5I0Vm79AX6IenI:b-4bkkiw) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/22e3e84e17d76d240d9ca0167261d7f48f92621d/stwo_cairo_verifier/src/queries.cairo#L68)):*
> ```Cairo
> 
> 
> pub fn get_sparse_sub_circle_domain_dict(
> ```

Where is this used?


<!-- Sent from Reviewable.io -->

atgrosso commented:
> > Reviewed all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101)_ status: 0 of 11 files reviewed, 4 unresolved discussions (waiting on @atgrosso)
> 
> _[`stwo_cairo_verifier/src/queries.cairo` line 30 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101#-O8EyLf69LBWKNM6WM_9:-O8EyLf69LBWKNM6WM_A:b-3dufzo) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/22e3e84e17d76d240d9ca0167261d7f48f92621d/stwo_cairo_verifier/src/queries.cairo#L30)):_
> 
> > ```cairo
> >             i = i + 1;
> >         };
> >         res
> > ```
> 
> I think its better for readability using for loops vs while loops where appropriate.
> 
> _Suggestion:_
> 
> ```cairo
>         let sub_circle_size = pow(2, *self.log_size);
>         let start = *self.coset_index * sub_circle_size;
>         let end = start + sub_circle_size;
>         let mut res = array![];
>         for i in start..end {
>             res.append(i);
>         };
>         res
> ```
> 
> _[`stwo_cairo_verifier/src/queries.cairo` line 51 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101#-O8F-xIUB2txPuljeEnG:-O8F-xIUB2txPuljeEnH:bj3lh64) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/22e3e84e17d76d240d9ca0167261d7f48f92621d/stwo_cairo_verifier/src/queries.cairo#L51)):_
> 
> > ```cairo
> > #[generate_trait]
> > pub impl SparseSubCircleDomainImpl of SparseSubCircleDomainTrait {
> >     fn flatten(self: @SparseSubCircleDomain) -> Span<usize> {
> > ```
> 
> Why not return Array?
> 
> _Code quote:_
> 
> ```cairo
> Span<usize>
> ```
> 
I don't remember the arguments of this decision and i don't see any downside using an array. So, i changed it to array
> _[`stwo_cairo_verifier/src/queries.cairo` line 63 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101#-O8F-zpW2xkqLcoXP7az:-O8F-zpW2xkqLcoXP7b-:b-t6gsf6) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/22e3e84e17d76d240d9ca0167261d7f48f92621d/stwo_cairo_verifier/src/queries.cairo#L63)):_
> 
> > ```cairo
> >             i = i + 1;
> >         };
> >         res.span()
> > ```
> 
> Same here perhaps
> 
> _Suggestion:_
> 
> ```cairo
>         let mut res = array![];
>         for domain in self.domains.span() {
>             for position in domain.to_decommitment_positions() {
>                 res.append(position);
>             };
>         };
>         res
> ```
> 
Same as the above
> _[`stwo_cairo_verifier/src/queries.cairo` line 68 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101#-O8Exk5I0Vm79AX6IenH:-O8Exk5I0Vm79AX6IenI:b-4bkkiw) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/22e3e84e17d76d240d9ca0167261d7f48f92621d/stwo_cairo_verifier/src/queries.cairo#L68)):_
> 
> > ```cairo
> > pub fn get_sparse_sub_circle_domain_dict(
> > ```
> 
> Where is this used?

Removed. It is used by the pcs verifier. I will include it then

andrewmilson reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/101#-:-O8wbW8SA_QzQEHCerul:bnfp4nl)* status: 0 of 11 files reviewed, 4 unresolved discussions (waiting on @atgrosso)



<!-- Sent from Reviewable.io -->
